lib/model/message_list.dart

```diff
@@ -35,14 +50,42 @@ mixin _MessageSequence {
   /// It exists as an optimization, to memoize the work of parsing.
   final List<ZulipContent> contents = [];
 
+  /// The messages and their siblings in the UI, in order.
+  ///
+  /// This has a [MessageListMessageItem] corresponding to each element
+  /// of [messages], in order.  It may have additional items interspersed
+  /// before, between, or after the messages.
+  ///
+  /// This information is completely derived from [messages].
+  /// It exists as an optimization, to memoize that computation.
+  final List<MessageListItem> items = [];
+
   int _findMessageWithId(int messageId) {
     return binarySearchByKey(messages, messageId,
       (message, messageId) => message.id.compareTo(messageId));
   }
 
+  int _findItemWithMessageId(int messageId) {
+    return binarySearchByKey(items, messageId, _compareItemToMessageId);
+  }
```

Comment:
Good question. It'll need some adjustments and I suppose some renamings for clarification, but I think the core logic will actually continue to work fine: the key is that
* it's really looking for the *message* item for the message ID;
* there'll still be exactly one of those per message;
* and the other items with a given message ID will all be in a predictable direction relative to the message item (specifically, date separator and recipient header will be before the message, and if we later dream up some item that goes after its associated message then that'll fit into this scheme too).

So when the comparison function sees some item that has the right message ID but isn't a message item, it can just say those compare before the key.